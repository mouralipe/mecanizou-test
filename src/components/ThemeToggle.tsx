'use client';

import { useTheme } from '@/hooks/useTheme';

export function ThemeToggle() {
  const { theme, toggleTheme } = useTheme();

  const themeIcon = theme === 'light' ? 'ðŸŒ™' : 'ðŸŒž';

  return (
    <button
      onClick={toggleTheme}
      aria-label={`Mudar para tema ${theme === 'light' ? 'escuro' : 'claro'}`}
    >
      {themeIcon}
    </button>
  );
}
