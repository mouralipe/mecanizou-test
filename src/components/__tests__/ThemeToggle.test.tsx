import { fireEvent, render, screen } from '@testing-library/react';
import { ThemeToggle } from '../ThemeToggle';

const mockToggleTheme = jest.fn();
const mockUseTheme = jest.fn();

jest.mock('@/hooks/useTheme', () => ({
  useTheme: () => mockUseTheme(),
}));

describe('ThemeToggle Component', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders sun icon when theme is light', () => {
    mockUseTheme.mockReturnValue({
      theme: 'light',
      toggleTheme: mockToggleTheme,
    });

    render(<ThemeToggle />);

    const button = screen.getByRole('button');
    expect(button).toBeInTheDocument();
    expect(button).toHaveTextContent('ðŸŒ™');
    expect(button).toHaveAttribute('aria-label', 'Mudar para tema escuro');
  });

  it('renders moon icon when theme is dark', () => {
    mockUseTheme.mockReturnValue({
      theme: 'dark',
      toggleTheme: mockToggleTheme,
    });

    render(<ThemeToggle />);

    const button = screen.getByRole('button');
    expect(button).toBeInTheDocument();
    expect(button).toHaveTextContent('ðŸŒž');
    expect(button).toHaveAttribute('aria-label', 'Mudar para tema claro');
  });

  it('calls toggleTheme when clicked', () => {
    mockUseTheme.mockReturnValue({
      theme: 'light',
      toggleTheme: mockToggleTheme,
    });

    render(<ThemeToggle />);

    const button = screen.getByRole('button');
    fireEvent.click(button);

    expect(mockToggleTheme).toHaveBeenCalledTimes(1);
  });

  it('handles multiple clicks correctly', () => {
    mockUseTheme.mockReturnValue({
      theme: 'dark',
      toggleTheme: mockToggleTheme,
    });

    render(<ThemeToggle />);

    const button = screen.getByRole('button');
    fireEvent.click(button);
    fireEvent.click(button);
    fireEvent.click(button);

    expect(mockToggleTheme).toHaveBeenCalledTimes(3);
  });

  it('has correct accessibility attributes', () => {
    mockUseTheme.mockReturnValue({
      theme: 'light',
      toggleTheme: mockToggleTheme,
    });

    render(<ThemeToggle />);

    const button = screen.getByRole('button');
    expect(button).toHaveAttribute('aria-label');
    expect(button.getAttribute('aria-label')).toContain('tema');
  });

  it('uses icon variant for button', () => {
    mockUseTheme.mockReturnValue({
      theme: 'light',
      toggleTheme: mockToggleTheme,
    });

    render(<ThemeToggle />);

    const button = screen.getByRole('button');
    expect(button).toHaveClass('cursor-pointer');
  });

  it('matches snapshot with light theme', () => {
    mockUseTheme.mockReturnValue({
      theme: 'light',
      toggleTheme: mockToggleTheme,
    });

    const { container } = render(<ThemeToggle />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches snapshot with dark theme', () => {
    mockUseTheme.mockReturnValue({
      theme: 'dark',
      toggleTheme: mockToggleTheme,
    });

    const { container } = render(<ThemeToggle />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('updates icon when theme changes', () => {
    const { rerender } = render(<ThemeToggle />);

    mockUseTheme.mockReturnValue({
      theme: 'light',
      toggleTheme: mockToggleTheme,
    });
    rerender(<ThemeToggle />);

    expect(screen.getByRole('button')).toHaveTextContent('ðŸŒ™');

    mockUseTheme.mockReturnValue({
      theme: 'dark',
      toggleTheme: mockToggleTheme,
    });
    rerender(<ThemeToggle />);

    expect(screen.getByRole('button')).toHaveTextContent('ðŸŒž');
  });
});
