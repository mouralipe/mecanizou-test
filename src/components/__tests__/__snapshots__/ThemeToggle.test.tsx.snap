// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`ThemeToggle Component matches snapshot with dark theme 1`] = `
<button
  aria-label="Mudar para tema claro"
  class="cursor-pointer"
>
  ðŸŒž
</button>
`;

exports[`ThemeToggle Component matches snapshot with light theme 1`] = `
<button
  aria-label="Mudar para tema escuro"
  class="cursor-pointer"
>
  ðŸŒ™
</button>
`;
